### PR TITLE
MCR-2966 move saxon compatible stylesheets to folder xslt

### PR DESCRIPTION
**Pull request by michael-becker**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2966).

The following changes were made:
* moved all XSL3 stylesheets (ending with -3.xsl) to the xslt folder
* created a new MCR configuration variable MCR.Layout.Transformer.Factory.XSLFolder=xslt
* replaced "xsl/"+stylsesheet statements in code with the configuration variable
* created a new URI resolver for XEditor templates: xslStyleXEditor and used this resolver in the xed
* created new iview2-solr.xsl in xslt folder and added isFileSupported to MCRIview2URIResolver

### CODE REVIEW

File: mycore-solr/src/main/resources/xslt/solr/indexing/mycoreobject-dynamicfields.xsl [hunk 1]
```diff
@@ -5,9 +5,9 @@
   xmlns:fn="http://www.w3.org/2005/xpath-functions"
   xmlns:mcrsolr="http://www.mycore.de/xslt/solr">
   
-  <xsl:import href="xslImport:solr-document-3:solr/indexing/mycoreobject-dynamicfields-3.xsl" />
+  <xsl:import href="xslImport:solr-document:solr/indexing/mycoreobject-dynamicfields.xsl" />
```
Comment (rsteph-de) [inline mycore-solr/src/main/resources/xslt/solr/indexing/mycoreobject-dynamicfields.xsl:8]: I think, the path should also start with `/xsl` or `/xslt` to distinguish version 1 and version 3 XSL templates.
Or does that break to much existing code?

File: mycore-solr/src/main/resources/components/solr/config/mycore.properties [hunk 2]
```diff
@@ -88,10 +73,8 @@ MCR.Solr.TikaFields=stream_size,stream_name,stream_source_info,stream_content_ty
 MCR.Solr.JoinQueryFields=%MCR.Solr.ContentBaseFields%,%MCR.Solr.AVExtenderFields%,%MCR.Solr.TikaFields%
 
 MCR.CLI.Classes.Internal=%MCR.CLI.Classes.Internal%,org.mycore.solr.commands.MCRSolrCommands
-MCR.URIResolver.xslImports.solr-document=%MCR.URIResolver.xslImports.solr-document%,mycoreobject-dynamicfields.xsl
-MCR.URIResolver.xslImports.solr-document-3=%MCR.URIResolver.xslImports.solr-document-3%,solr/indexing/mycoreobject-dynamicfields-3.xsl
-MCR.URIResolver.xslIncludes.xeditorTemplates=%MCR.URIResolver.xslIncludes.xeditorTemplates%,solr-xeditor-templates.xsl
-MCR.URIResolver.xslIncludes.xeditorTemplates-3=%MCR.URIResolver.xslIncludes.xeditorTemplates-3%,solr/xeditor/solr-xeditor-templates-3.xsl
+MCR.URIResolver.xslImports.solr-document=%MCR.URIResolver.xslImports.solr-document%,solr/indexing/mycoreobject-dynamicfields.xsl
+MCR.URIResolver.xslIncludes.xeditorTemplates=%MCR.URIResolver.xslIncludes.xeditorTemplates%,solr/xeditor/solr-xeditor-templates.xsl
```
Comment (rsteph-de) [inline mycore-solr/src/main/resources/components/solr/config/mycore.properties:79]: I think, the path should also start with `/xsl` or `/xslt` to distinguish version 1 and version 3 XSL templates.
Or does that break to much existing code?

**Review by rsteph-de — CHANGES_REQUESTED** [review]
Bei den Imports verzichten wir noch beim Pfadanfang auf "/xslt" bzw. "/xsl"?
Auch wenn dadurch Dinge "kaputt" gehen, würde ich das gerne vereinheitlichen.
Die "Magie" sollte noch raus. Außerdem können wir so besser erkennen, ob es sich um alte oder neue Stylesheets handelt.

**Comment by michael-becker** [comment]
> Bei den Imports verzichten wir noch beim Pfadanfang auf "/xslt" bzw. "/xsl"? Auch wenn dadurch Dinge "kaputt" gehen, würde ich das gerne vereinheitlichen. Die "Magie" sollte noch raus. Außerdem können wir so besser erkennen, ob es sich um alte oder neue Stylesheets handelt.

Ja wäre ich auch dafür. Das haben wir momentan noch raus gelassen, um den Umstieg etwas sanfter zu machen. Nehmen wir die vollständige Angabe mit resource:xslt/... müssen in allen Anwendungen, die 2023 und XSL1 verwenden, noch mehr an der Konfiguration angepasst werden als ohnehin schon - siehe die Mail, du warst zu schnell ;)

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java [hunk 3]
```diff
@@ -1357,7 +1376,30 @@ private static class MCRXslImportResolver implements URIResolver {
 
         @Override
         public Source resolve(String href, String base) throws TransformerException {
-            String importXSL = MCRXMLFunctions.nextImportStep(href.substring(href.indexOf(':') + 1));
+            final String baseURI = getParentDirectoryResourceURI(base);
+            // set xslt folder
+            final String xslFolder;
+            if (StringUtils.contains(baseURI, "/xsl/")) {
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: baseURI.startswith would work more secure

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java [hunk 4]
```diff
@@ -1357,7 +1376,30 @@ private static class MCRXslImportResolver implements URIResolver {
...
+            // check for old import format: xslImport:property:current.xsl and issue a warning
+            final String importXSL;
+            final String importPart = StringUtils.substringAfter(href, ":");
+            if (StringUtils.contains(importPart, ":")) {
+                final String propertyPart = StringUtils.substringBefore(importPart, ":");
+                LOGGER.warn("{} is in old import format change to xslImport:{}!", href, propertyPart);
+
+                importXSL = MCRXMLFunctions.nextImportStep(importPart);
+            } else {
+                final String selfName = StringUtils.substringAfter(baseURI, xslFolder + "/")
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: maybe more secure to use baseURI.substring(xslFolder.length+1)

File: mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java (hunk 3 above)
Comment (michael-becker) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: If I understand this correctly, getParentDirectoryResourceURI always returns a string starting with resource (https://github.com/MyCoRe-Org/mycore/blob/2022.06.x/mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java#L293)
e.g.:
base: jar:file:/home/musiclpz/production/build/webapps/WEB-INF/lib/mycore-solr-2023.08-SNAPSHOT.jar!/xslt/solr/indexing/mycoreobject-dynamicfields.xsl
baseuri: resource:xslt/solr/indexing/

so startsWith resource:xsl/ should be feasible.

However, in some edge cases getParentDirectoryResourceURI returns the given base. In these cases, this would not work. But I am not totally aware if in these cases something else goes wrong. If we can ignore these edge cases, we can test on startsWith resource:xsl/

Comment (sebhofmann) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: > If I understand this correctly, getParentDirectoryResourceURI always returns a string starting with resource (https://github.com/MyCoRe-Org/mycore/blob/2022.06.x/mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java#L293)
e.g.:
base: jar:file:/home/musiclpz/production/build/webapps/WEB-INF/lib/mycore-solr-2023.08-SNAPSHOT.jar!/xslt/solr/indexing/mycoreobject-dynamicfields.xsl
baseuri: resource:xslt/solr/indexing/

Yes you are right.

> However, in some edge cases getParentDirectoryResourceURI returns the given base. In these cases, this would not work. But I am not totally aware if in these cases something else goes wrong. If we can ignore these edge cases, we can test on startsWith resource:xsl/

I think in the edge cases the property should be used. So check for startsWith resource:xsl/ or resource:xslt/ and if both false than use the property.

Comment (michael-becker) [reply mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java]: Changed to startsWith resource:xsl/ / resource:xslt/

I had to rollback the changes in nextImportStep, since it produced errors when using the proposed new xslImport syntax (without selfname). This resulted in errors because xsl:import href="xslImport:xxx" was included more than once and ultimately failed with a stylesheet including itself. I was not able to fix this and, thus. would suggest to postpone the syntax change to a later point.

File: mycore-solr/src/main/resources/xslt/solr/indexing/mycoreobject-dynamicfields.xsl (hunk 1 above)
Comment (michael-becker) [reply mycore-solr/src/main/resources/xslt/solr/indexing/mycoreobject-dynamicfields.xsl:8]: Currently, we cann not use resource:xslt/ in xslImport, since the respective URI resolver always adds a resource: on its own (https://github.com/MyCoRe-Org/mycore/blob/issues/MCR-2966-xslt-folder/mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java#L1400)

File: mycore-solr/src/main/resources/components/solr/config/mycore.properties (hunk 2 above)
Comment (michael-becker) [reply mycore-solr/src/main/resources/components/solr/config/mycore.properties:79]: as abvoe: xlsImport does not allow for adding resource:xsl/ or /xslt/ in the beginning. I would suggest to adapt these references in a separate PR after we added this.

**Comment by michael-becker** [comment]
Falls weiterhin XSLT-1 verwendet werden soll, sind folgende Anpassungen an der Konfiguration vorzunehmen:
* Für alle Anwendungen
  * MCR.Layout.Transformer.Factory.XSLFolder=xsl
  * MCR.LayoutService.TransformerFactoryClass=org.apache.xalan.processor.TransformerFactoryImpl

* Anwendungen, die mycore-solr nutzen
  * MCR.URIResolver.xslImports.solr-document=solr-basetemplate.xsl
  * MCR.URIResolver.xslImports.solr-document=%MCR.URIResolver.xslImports.solr-document%,mycoreobject-dynamicfields.xsl
  * MCR.URIResolver.xslIncludes.components=%MCR.URIResolver.xslIncludes.components%,solr-layout-utils.xsl
  * MCR.ContentTransformer.response.TransformerFactoryClass=org.apache.xalan.processor.TransformerFactoryImpl
  * MCR.ContentTransformer.response.Stylesheet=%MCR.ContentTransformer.response-prepared.Stylesheet%,xsl/response.xsl
  * MCR.ContentTransformer.response-browse.Stylesheet=xsl/response-browse.xsl
  * MCR.ContentTransformer.mycoreobject-solrdocument.Class=org.mycore.common.content.transformer.MCRXSL2JAXBTransformer
  * MCR.ContentTransformer.mycoreobject-solrdocument.TransformerFactoryClass=org.apache.xalan.processor.TransformerFactoryImpl
  * MCR.ContentTransformer.mycoreobject-solrdocument.Stylesheet=xsl/mycoreobject-solrdocument.xsl
  * MCR.ContentTransformer.mycoreobject-solrdocument.Context=org.mycore.solr.index.document.jaxb
  * MCR.ContentTransformer.response-solrdocument.Class=org.mycore.common.content.transformer.MCRXSL2JAXBTransformer
  * MCR.ContentTransformer.response-solrdocument.TransformerFactoryClass=org.apache.xalan.processor.TransformerFactoryImpl
  * MCR.ContentTransformer.response-solrdocument.Stylesheet=xsl/response2batch.xsl,%MCR.ContentTransformer.mycoreobject-solrdocument.Stylesheet%
  * MCR.ContentTransformer.response-solrdocument.Context=%MCR.ContentTransformer.mycoreobject-solrdocument.Context%
  * MCR.ContentTransformer.response-prepared.Class=org.mycore.common.content.transformer.MCRXSLTransformer
  * MCR.ContentTransformer.response-prepared.TransformerFactoryClass=org.apache.xalan.processor.TransformerFactoryImpl
  * MCR.ContentTransformer.response-prepared.Stylesheet=xsl/response-join-results.xsl,xsl/response-addDocId.xsl,xsl/response-addDerivates.xsl
  * MCR.URIResolver.xslIncludes.xeditorTemplates=solr-xeditor-templates.xsl

* Anwendungen, die mycore-mods nutzen
  * MCR.URIResolver.xslImports.solr-document=%MCR.URIResolver.xslImports.solr-document%,mods-solr.xsl

Anwendungen, die schon auf XSLT3 umgestellt sind, brauchen folgende Einträge:
* MCR.LayoutService.TransformerFactoryClass=net.sf.saxon.TransformerFactoryImpl
* MCR.URIResolver.xslIncludes.xeditorTemplates=resource:xsl/solr-xeditor-templates.xsl

**Review by yagee-de — APPROVED** [review]
Works with MIR branch https://github.com/MyCoRe-Org/mir/pull/913
